File: nancykim99/boj/c++/boj10799.cpp
/*
BOJ10799 : 쇠막대기 (S2)

해결 방법 : 레이저가 나오면, 파이프만큼 개수를 늘리고, 파이프가 끝나면 파이프 개수를 줄이는 방식으로 해결.

*/

#include<iostream>
#include<stack>
#include<string.h>
using namespace std;
int main()
{
	stack<int> stick;
	string str;
	
	int cnt = 0;
	cin >> str;
	for (int i = 0; i < str.size(); i++)
	{
		if (str[i] == '(' && str[i+1] == ')')
		{
			cnt += stick.size();
			i++;
		}
		else if (str[i] == '(')
		{
			stick.push(i);
		}
		else if (str[i] == ')')
		{
			cnt++;
			stick.pop();
		}
	}
	cout << cnt;
}

File: yongha_swea/BOJ/boj10799.py
#BOJ10799 쇠막대기

sticks = input()

stick_stack = []

count = 0

for i in range(len(sticks)):
    if sticks[i] == '(':
        stick_stack.append('(')
    else:
        # 레이저 발싸 ('()') - stack에 들어간 갯수 = 막대기 높이, 전체 다 절단되기 때문에 그 수 만큼 카운트 추가
        if sticks[i-1] == '(':
            stick_stack.pop()
            count += len(stick_stack)
        # 막대 끝이라는 뜻, 막대 한 조각을 count에 더해주기
        else:
            stick_stack.pop()
            count += 1

#전체 조각수 출력
print(count)